### PR TITLE
Add new log field to output server name sent by client in TLS handshake

### PR DESCRIPTION
**Pull request by ywkaras**

### CODE REVIEW

**Comment by shinrich** [comment]
[approve ci]

**Comment by ywkaras** [comment]
[approve ci]

**Comment by shinrich** [comment]
Looks good to me.  I'd like to get someone outside of Yahoo to take a look too though.

**Comment by maskit** [comment]
This is not only about this PR, but what do we want to do for QUIC connections?

If this is just for ordinary HTTPS connections (and we'll have another log field for QUIC), I'm fine with this. If not, since this is a new feature, we should have implementation for QUIC connections too or should file an issue as todo.

Also, this is up to RM but we should update Release Note if the change is not going into 9.0.

**Comment by ywkaras** [comment]
Hmm I don't see a server name in MockNetVConnection:

https://github.com/apache/trafficserver/blob/c7a4ffafd8eeb0e4b80ed057963722c33029acc8/iocore/net/quic/Mock.h#L214

**Comment by maskit** [comment]
I don't think supporting QUIC is that easy, and that's a reason I left an option to file it as a todo. But if possible, adding the same function to QUICNetVC wouldn't be a bad idea even if the function was a stub.

Because of the messy inheritance relationship between UnixNetVC and SSLNetVC we couldn't have a parent class for SSLNetVC and QUICNetVC, but we might be able to do something similar to ALPNSupport class.

**Comment by ywkaras** [comment]
Does QUIC have a mechanism similar to SNI?  Where one end has multiple public keys, and it decides which one to provide based on an identifier sent from the other end?

**Comment by maskit** [comment]
Yes, QUIC itself doesn't mention SNI but it's a part of TLS handshake ([QUIC uses TLS](https://tools.ietf.org/html/draft-ietf-quic-tls-23)). We should be able to do the same thing. However, I'm not sure if current our QUIC implementation supports it.

File: proxy/logging/LogAccess.cc [hunk 1]
```diff
@@ -1264,6 +1264,26 @@ LogAccess::marshal_cache_lookup_url_canon(char *buf)
...
 
+/*-------------------------------------------------------------------------
+  -------------------------------------------------------------------------*/
+int
+LogAccess::marshal_client_sni_server_name(char *buf)
+{
+  // NOTE:  For this string_view, data() must always be nul-terminated, but the nul character must not be included in
+  // the length.
+  //
+  std::string_view sn = "n/a";
```
Comment (randall) [inline proxy/logging/LogAccess.cc]: returning `n/a` is non-standard for ATS. The default for no value ends up being `-` (which I think happens if you don't set anything in `buf`)

**Comment by ywkaras** [comment]
@randall OK fixed.

**Comment by randall** [comment]
@maskit Can this be merged as is? Or do we need additional changes for QUIC?

**Comment by maskit** [comment]
@randall To support this on QUIC connection additional changes will be needed, but we can merge this as is. In that case I want to see an issue for QUIC as a todo (missing feature).

**Comment by maskit** [comment]
Oh, but be careful with the release note. We should make sure that this can go into 9.0 if we don't have release note change on this PR.

**Comment by ywkaras** [comment]
What's the rule of thumb for release notes?  Should all changes that are not purely refactoring correspond to some release note?

**Comment by maskit** [comment]
> What's the rule of thumb for release notes? Should all changes that are not purely refactoring correspond to some release note?

I would not say all changes, but this one should be because it's clearly new feature.

@zwoop What's the plan? We need a guideline somewhere.

**Comment by ywkaras** [comment]
Be advised I'll have to ask Vz committers to -1 any 9.x release unless this is in it, as it's important to Vz ops people.

**Comment by maskit** [comment]
@ywkaras Did you create an issue for QUIC? Or are you working on it?

**Comment by ywkaras** [comment]
@maskit  OK https://github.com/apache/trafficserver/issues/6050

**Comment by bryancall** [comment]
@ywkaras 
As the code stands with std::string in the HttpSM we are going to have memory leaks.  You should be able to see this with running an ASAN build.

We should move vc information that needs to be around for logging into the session, since it will be around at the time we log.  We should create a struct inside of the Session (say ClientLogInfo or something better) that is used for logging.  This way we only allocate and copy this information 1 time.  The HttpSM should call getters on ClientLogInfo to log this information.

**Comment by SolidWallOfCode** [comment]
Given this code, that `std::string` is going to leak memory, presuming it doesn't crash.

**Comment by SolidWallOfCode** [comment]
_Will_ the session be around at the time of transaction logging? I'd like @shinrich to weigh in on that. If so, then Bryan's right, this should be moved to the client session object. As a temporary expedient this could have a `string_view` in the `HttpSM` and using the `Arena` in `HttpTransact` to store the data - this avoids `malloc` almost all the time and doesn't have cleanup issues.

**Comment by ywkaras** [comment]
Why would this cause a memory leak?  Do we somehow bypass calling ~HttpSM() ?

**Comment by ywkaras** [comment]
@zwoop what is the better way to save a variable-length server name for later logging, than with std::string?

**Comment by ywkaras** [comment]
@SolidWallOfCode No the session object is gone by the time the log entry is generated.  You can see preexisting code to save stuff from the session object, for that reason.

**Comment by ywkaras** [comment]
Are we needing a little C++ 101 here?  The destructors of data members are implicitly called by the (implicit or explicit) class destructor:  https://godbolt.org/z/wyrNSn .  See lines 9 and 20 of the assembler output.

**Comment by SolidWallOfCode** [comment]
`HttpSM::~HttpSM` is never called.

If the session object is gone, then I'd go with using the `Arena`.

**Comment by ywkaras** [comment]
Thanks for catching that @zwoop .

**Comment by SolidWallOfCode** [comment]
[approve ci]

**Comment by bryancall** [comment]
@ywkaras I would just use a const char* inside of the session and do a `strdup`

**Comment by ywkaras** [comment]
@bryancall The session object is gone when the log entry to generated.  @SolidWallOfCode wanted it in the arena.

**Comment by ywkaras** [comment]
@zwoop the std::string is KWF.

**Pull request closed without merging**